server/src/main/java/com/flipkart/varadhi/qos/RateLimiter.java

```diff
@@ -0,0 +1,228 @@
+
+import java.util.*;
+import java.util.concurrent.*;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+
+/**
+ * - Initialise a priorityQueue (min heap) to maintain (time, order, data).
+ * - Custom comparator to sort by time asc, order desc.
+ *
+ * Thread 1:
+ * - pods will send data (from, to, data), break it into 2 events.
+ * - (from, 1, +data/(to-from)) and (to, 0, -data/(to-from)) and add into PQ if
+ * (from > (now() - Window size)) // skip adding if old data
+ *
+ * Thread 2: (pooling)
+ * - maintain aggreagte value to use in check breach.
+ * - peek and pop till now()
+ * - compute running avg.
+ * - check for RL. If true, add the suppression factor
+ */
+
+@Slf4j
+public class RateLimiter { // TODO(rl): impl publisher, for observer
+
+    private final int windowSize;
+    private final ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(1);
+    private final VaradhiTopicService varadhiTopicService;
+    private final SuppressionData<Float> suppressionData;
+    private final Multiset<LoadInfo> loadHistory;
+    private final WeightFunction weightFunction;
+    private final List<RLObserver> observers = new ArrayList<>();
+
+    public RateLimiter(VaradhiTopicService varadhiTopicService, int windowSize, WeightFunction weightFunction) {
+        this.varadhiTopicService = varadhiTopicService;
+        this.windowSize = windowSize;
+        this.weightFunction = weightFunction;
+        this.loadHistory = TreeMultiset.create();
+        this.suppressionData = new SuppressionData<>();
+        sendSuppressionFactorToWeb();
+    }
+
+    public void addTrafficData(LoadInfo dump) {
+        loadHistory.add(dump);
+        log.debug("received data: " + dump.getTopicUsageMap());
+        log.debug("current loadHistory: " + loadHistory);
+    }
+
+    // TODO(rl): add try block in all scheduledServices
+    public void sendSuppressionFactorToWeb() {
+        scheduledExecutorService.scheduleAtFixedRate(() -> {
+            try {
+                log.debug("Update suppression factor");
+                var trafficHistoryMap = generateDS();
+                log.debug("generated trafficHistoryMap: " + trafficHistoryMap);
+                sendSuppressionFactor(trafficHistoryMap);
+            } catch (Exception e) {
+                log.error("Exception occurred while sending suppression factor", e);
+            }
+        }, 0, windowSize, TimeUnit.SECONDS);
+    }
+
+    private Map<String, Multiset<Pair<Long, TrafficData>>> generateDS() {
+        // takes loadHistory and generate trafficHistoryMap;
+        Map<String, Multiset<Pair<Long, TrafficData>>> trafficHistoryMap = new HashMap<>();
+
+        // remove old values (older than windowSize)
+        long popTill = System.currentTimeMillis() - (windowSize * 1000L);
+        loadHistory.removeIf(loadInfo -> loadInfo.getTo() < popTill);
```

Comment:
looks expensive.

Reply:
not needed now, maintaining a deque that will handle this operation in optimal way.